File: FamilyFinance/Services/CategoryService.cs
using FamilyFinance.Exceptions;
using FamilyFinance.Models.Dtos;
using FamilyFinance.Models.Entities;
using FamilyFinance.Models.Requests;
using FamilyFinance.Models.Requests.ListFilters;
using FamilyFinance.Models.Responses;
using FamilyFinance.Models.Responses.ListItems;
using FamilyFinance.Utils;
using Microsoft.EntityFrameworkCore;

namespace FamilyFinance.Services;

public class CategoryService(AppDbContext dbContext) : BaseService(dbContext)
{
    /// <summary>
    /// Get a list of categories with pagination and optional search filtering.
    /// </summary>
    /// <param name="filter"></param>
    /// <returns></returns>
    public async Task<Paginated<CategoryListItem>> ListAsync(CategoryListFilter filter)
    {
        var query = DbContext.Categories.AsQueryable();
        if (!string.IsNullOrEmpty(filter.SearchTerm))
        {
            var keyword = filter.SearchTerm.ToLower();
            query = query.Where(c => c.Name.ToLower().Contains(keyword) || c.Description.ToLower().Contains(keyword) || c.Color.ToLower().Contains(keyword));
        }

        var totalCount = await query.CountAsync();
        var items = await query
            .OrderBy(c => c.Name)
            .Skip((filter.Page - 1) * filter.PageSize)
            .Take(filter.PageSize)
            .Select(c => new CategoryListItem(c))
            .ToListAsync();

        return new Paginated<CategoryListItem> {
            Items = items,
            TotalCount = totalCount,
            Page = filter.Page,
            PageSize = filter.PageSize
        };
    }

    /// <summary>
    /// Get a category by its ID.
    /// </summary>
    /// <param name="id"></param>
    /// <returns></returns>
    public async Task<CategoryDto?> GetByIdAsync(Guid id)
    {
        var category = await DbContext.Categories
            .FirstOrDefaultAsync(c => c.Id == id);
        return category != null ? new CategoryDto(category) : null;
    }

    /// <summary>
    /// Create a new category.
    /// </summary>
    /// <param name="request"></param>
    /// <returns></returns>
    public async Task<CategoryDto> CreateAsync(CategorySaveRequest request)
    {
        var newCategory = new Category {
            Name = request.Name,
            Description = request.Description,
            Color = request.Color,
            CreatedAt = request.Timestamp ?? DateTime.Now,
            CreatedBy = request.UserId ?? throw new UnauthorizedAccessException(),
        };
        await DbContext.Categories.AddAsync(newCategory);
        await DbContext.SaveChangesAsync();
        return new CategoryDto(newCategory);
    }

    /// <summary>
    /// Update an existing category by its ID.
    /// </summary>
    /// <param name="id"></param>
    /// <param name="request"></param>
    /// <returns></returns>
    /// <exception cref="EntityNotFoundException"></exception>
    public async Task<CategoryDto> UpdateAsync(Guid id, CategorySaveRequest request)
    {
        var category = await DbContext.Categories.FindAsync(id)
            ?? throw new EntityNotFoundException($"Category with ID {id} not found.");

        category.Name = request.Name;
        category.Description = request.Description;
        category.Color = request.Color;
        DbContext.Categories.Update(category);
        await DbContext.SaveChangesAsync();
        return new CategoryDto(category);
    }

    /// <summary>
    /// Delete a category by its ID.
    /// </summary>
    /// <param name="id"></param>
    /// <returns></returns>
    /// <exception cref="EntityNotFoundException"></exception>
    public async Task DeleteAsync(Guid id)
    {
        var category = await DbContext.Categories.FindAsync(id)
            ?? throw new EntityNotFoundException($"Category with ID {id} not found.");
        DbContext.Categories.Remove(category);
        await DbContext.SaveChangesAsync();
    }
}

File: family_finance_api_go/internal/services/category_service.go
package services

import (
	"github.com/dimasyanu/family-finance-go/internal/models"
	"github.com/dimasyanu/family-finance-go/internal/models/request"
	"github.com/dimasyanu/family-finance-go/internal/models/request/filter"
	"github.com/dimasyanu/family-finance-go/internal/models/response"
	"github.com/dimasyanu/family-finance-go/internal/repositories"
)

type CategoryService struct {
	repo *repositories.CategoryRepository
}

func NewCategoryService(repo *repositories.CategoryRepository) *CategoryService {
	return &CategoryService{repo: repo}
}

func (cs *CategoryService) List(f *filter.CategoryListFilter) (*response.Paginated[models.Category], error) {
	items, total := cs.repo.List(f)
	return &response.Paginated[models.Category]{Items: items, Total: total}, nil
}

func (cs *CategoryService) GetByID(id uint) (*models.Category, error) {
	category, err := cs.repo.GetByID(id)
	if err != nil {
		return nil, err
	}
	return category, nil
}

func (cs *CategoryService) Create(payload *request.SaveCategoryRequest) (uint, error) {
	model := models.FromCreateCategoryRequest(payload)

	id, err := cs.repo.Create(model)
	if err != nil {
		return 0, err
	}

	return id, nil
}

func (cs *CategoryService) Update(id uint, payload *request.SaveCategoryRequest) (uint, error) {
	category, err := cs.repo.GetByID(id)
	if err != nil {
		return 0, err
	}

	category.Name = payload.Name
	category.Description = payload.Description
	category.Color = payload.Color
	category.Icon = payload.Icon
	category.UserID = payload.UserId

	err = cs.repo.Update(category)
	if err != nil {
		return 0, err
	}

	return id, nil
}

func (cs *CategoryService) Delete(id uint) error {
	err := cs.repo.Delete(id)
	return err
}
